src/System.Memory/src/System/Buffers/ArrayMemoryPool.ArrayMemoryPoolBuffer.cs

```diff
@@ -82,17 +83,23 @@ public sealed override void Retain()
                 if (IsDisposed)
                     ThrowHelper.ThrowObjectDisposedException_ArrayMemoryPoolBuffer();
 
-                _refCount++;
+                Interlocked.Increment(ref _refCount);
             }
 
             public sealed override bool Release()
             {
                 if (IsDisposed)
                     ThrowHelper.ThrowObjectDisposedException_ArrayMemoryPoolBuffer();
 
-                int newRefCount = --_refCount;
+                int newRefCount = Interlocked.Decrement(ref _refCount);
```

Comment:
If you actually want thread-safety, these Interlocked.Decrement/Increments should be changed to CompareExchange loops, e.g.
```C#
Seems like this needs to be a CompareExchange loop, e.g.
```C#
// in Retain
while (true)
{
    int currentCount = Volatile.Read(ref _refCount);
    if (currentCount <= 0) ThrowHelper.ThrowObjectDisposedException_ArrayMemoryPoolBuffer();
    if (Interlocked.CompareExchange(ref _refCount, currentCount + 1, currentCount) == currentCount) break;
}
...
// in Release
while (true)
{
    int currentCount = Volatile.Read(ref _refCount);
    if (currentCount <= 0) ThrowHelper.ThrowObjectDisposedException_ArrayMemoryPoolBuffer();
    if (Interlocked.CompareExchange(ref _refCount, currentCount - 1, currentCount) == currentCount)
    {
        if (currentCount == 1)
        {
            Dispose();
            return true;
        }
        return false;
    }
}
```
Otherwise, consider this:
1. The current _refCount is 1.
2. Thread A calls Retain and gets past the IsDisposed check.
3. Thread B calls Release, disposing the object, and drops the count to 0.
4. Thread A resumes, increments the _refCount, and proceeds to party on the array thinking it successfully owns it.
5. Lots of other threads come along, and all Retain/Release, and as the ref count is now positive again, they're all successful.

Reply:
@AtsushiKan could you look at adding thread safety tests here?